man/man1/cwb_mplugin.1
.\" Man page generated from reStructuredText.
.
.TH "CWB_MPLUGIN" "1" "Jan 14, 2019" "" "coherent WaveBurst"
.SH NAME
cwb_mplugin \- cWB Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.nf

.fi
.sp
.nf

\fBUsage\fP

This command is used to merge plugins into one.

\fBSyntax\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBcwb_mplugin out_plugin.C in_plugin_1.C(opt) in_plugin_2.C(opt) in_plugin_3.C(opt) in_plugin_4.C(opt)\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBout_plugin.C\fP : name of the output plugin
.IP \(bu 2
\fBin_plugin_1.C\fP : name of the input plugin 1
.IP \(bu 2
\fBin_plugin_2.C\fP : name of the input plugin 2 (optional)
.IP \(bu 2
\fBin_plugin_3.C\fP : name of the input plugin 3 (optional)
.IP \(bu 2
\fBin_plugin_4.C\fP : name of the input plugin 4 (optional)
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
the input plugins are executed following the order of the declaration
in_plugin_1.C, in_plugin_2.C, in_plugin_3.C, in_plugin_4.C
.IP \(bu 2
the instructions #define,#include,#pragma must be defined outside the functions
.UNINDENT
.UNINDENT
.UNINDENT
.nf

\fBExample\fP

.fi
.sp
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
cwb_mplugin out_plugin.C in_plugin_1.C in_plugin_2.C
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SH AUTHOR
Gabriele Vedovato
.SH COPYRIGHT
2018, cWB Team
.\" Generated by docutils manpage writer.
.


tools/cwb/macros/cwb_mplugin.C
/*
# Copyright (C) 2019 Gabriele Vedovato
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/


// merge plugins into one plugin
{

  #define nPLUGIN 6

  CWB::Toolbox TB;

  TB.checkFile(gSystem->Getenv("CWB_ROOTLOGON_FILE"));
  TB.checkFile(gSystem->Getenv("CWB_PARAMETERS_FILE"));
  TB.checkFile(gSystem->Getenv("CWB_NETC_FILE"));

  // get used defined plugins up to nPLUGIN plugins
  TString cwb_mplugin[nPLUGIN];
  TMacro  cwb_mplugin_macro[nPLUGIN];
  for(int i=0;i<nPLUGIN;i++) {
    char env_label[32];sprintf(env_label,"CWB_MPLUGIN_%d",i+1);
    cwb_mplugin[i]=TString(gSystem->Getenv(env_label));
    if(i==0) { 		// this is the output plugin
      if(cwb_mplugin[i].CompareTo("")!=0) {
        TString DirName  = gSystem->DirName(cwb_mplugin[0]);
        TString BaseName = gSystem->BaseName(cwb_mplugin[0]);
        TB.checkFile(DirName);
        bool overwrite=TB.checkFile(cwb_mplugin[i],true);
        if(!overwrite) gSystem->Exit(1);
      } else {
        cout << "cwb_mplugin.C error : output file name not declared !!!" << endl;
        gSystem->Exit(1);
      }
    } else {
      if(cwb_mplugin[i].CompareTo("")!=0) {
        TB.checkFile(cwb_mplugin[i]);
        cwb_mplugin_macro[i] = TMacro(cwb_mplugin[i]); 
      }
    }
  }

  // print plugin list
  cout << "Output Plugin Name : " << endl << endl; 
  cout << 0 << " -> : " << cwb_mplugin[0] << endl;
  cout << endl << "Input Plugin Names : " << endl << endl; 
  for(int i=1;i<nPLUGIN;i++) {
    if(cwb_mplugin[i].CompareTo("")!=0) {
      cout << i << " -> " << cwb_mplugin[i] << endl;
    }
  }
  cout << endl;
  char answer[256];
  strcpy(answer,"");
  do {
    cout << "Do you want to continue it ? (y/n) ";
    cin >> answer;
    cout << endl << endl;
  } while ((strcmp(answer,"y")!=0)&&(strcmp(answer,"n")!=0));
  if (strcmp(answer,"n")==0) gSystem->Exit(0);


  // write plugin, if plugin is defined then we merge cwb_inet_plugin & plugin
  ofstream out;
  out.open(cwb_mplugin[0].Data(),ios::out);
  if(!out.good()) {cout << "cwb_mplugin.C - Error : Opening File : " 
                        << cwb_mplugin[0] << endl;gSystem->Exit(1);}

  // write warning message
  out << "// -------------------------------------------------------------------------" << endl;
  out << "// WARNING, NOT EDIT : This is a multi plugin generated with cwb_mplugin !!!" << endl;
  out << "// NOTE              : The main is listed on the bottom                      " << endl;
  out << "// " << endl;
  out << "// INPUT PLUGINS     :                                                       " << endl;
  out << "// " << endl;
  for(int i=1;i<nPLUGIN;i++) {
    if(cwb_mplugin[i].CompareTo("")!=0) {
      out << "// " << i << " - " << gSystem->BaseName(cwb_mplugin[i]) << endl; 
    }
  }
  out << "// " << endl;
  out << "// -------------------------------------------------------------------------" << endl;

  // write user user plugin code
  for(int i=1;i<nPLUGIN;i++) {
    if(cwb_mplugin[i].CompareTo("")!=0) {
      bool include=true;
      char user_plugin_name[32];sprintf(user_plugin_name,"CWB_UserPlugin_%d(",i);
      char user_plugin_namespace[32];sprintf(user_plugin_namespace,"CWB_UserPluginNamespace_%d",i);
      TList* fLines = cwb_mplugin_macro[i].GetListOfLines();
      TObjString *obj;
      TIter next(fLines);
      out << endl << endl;
      out << "// -------------------------------------------------------------------------" << endl;
      out << "// --> BEGIN CWB_USER PLUGIN CODE " << i << endl;
      out << "// " << endl; 
      out << "// " << gSystem->BaseName(cwb_mplugin[i]) << endl; 
      out << "// -------------------------------------------------------------------------" << endl;
      out << endl;
      while ((obj = (TObjString*) next())) {
        TString line = obj->GetName();
        TString sline = line;
        sline.ReplaceAll(" ","");
        if(sline.BeginsWith("#include")||sline.BeginsWith("#define")||sline.BeginsWith("#pragma")) {
          if(!include) out << "}" << endl;
          include=true; 
        } else {
          if(sline!="") {
            if(include) out << "namespace " << user_plugin_namespace << " {" << endl << endl;
            include=false;
          }
        }
        out <<  line.Data() << endl;
      }
      out << "}" << endl << endl;
      out << "// -------------------------------------------------------------------------" << endl;
      out << "// --> END CWB_USER PLUGIN CODE "   << i << endl;
      out << "// " << endl; 
      out << "// " << gSystem->BaseName(cwb_mplugin[i]) << endl; 
      out << "// -------------------------------------------------------------------------" << endl;
    }
  }

  // write user plugin declarations
  out << endl << endl;
  out << "// -------------------------------------------------------------------------" << endl;
  out << "// --> MAIN CWB_USER PLUGIN CODE " << endl;
  out << "// " << endl;
  out << "// INPUT PLUGINS     :                                                       " << endl;
  out << "// " << endl;
  for(int i=1;i<nPLUGIN;i++) {
    if(cwb_mplugin[i].CompareTo("")!=0) {
      out << "// " << i << " - " << gSystem->BaseName(cwb_mplugin[i]) << endl; 
    }
  }
  out << "// " << endl;
  out << "// -------------------------------------------------------------------------" << endl;
  out << endl;
  out << "#define XIFO 4" << endl;
  out << "#pragma GCC system_header" << endl;
  out << "#include \"cwb.hh\"" << endl;

  // write mplugin main
  out << endl << endl;
  out << "void" << endl;
  out << "CWB_Plugin(TFile* jfile, CWB::config* cfg, network* net, WSeries<double>* x, TString ifo, int type)  {" << endl;
  out << endl;
  for(int i=1;i<nPLUGIN;i++) {
    if(cwb_mplugin[i].CompareTo("")!=0) {
      char user_plugin_namespace[32];sprintf(user_plugin_namespace,"CWB_UserPluginNamespace_%d",i);
      out << "  " << user_plugin_namespace;
      out << "::CWB_Plugin(jfile, cfg, net, x, ifo, type); // CALL USER PLUGIN CODE " << i << endl;
    }
  }
  out << "}" << endl;

  out.close();

  cout << "Output Plugin Name : " << cwb_mplugin[0] << endl << endl;

  gSystem->Exit(0);
}
